batch2/batch/driver/instance_pool.py

```diff
@@ -287,20 +287,24 @@ def add_instance(self, instance):
                 return
             raise
 
-    async def handle_preempt_event(self, instance):
+    async def handle_preempt_event(self, instance, timestamp):
+        await instance.mark_jobs_ended(timestamp)
         await self.call_delete_instance(instance)
 
-    async def handle_delete_done_event(self, instance):
+    async def handle_delete_done_event(self, instance, timestamp):
+        await instance.mark_jobs_ended(timestamp)
```

Comment:
I think this should only run when the instance is stopped or deleted.  As far as I can tell, Google charges us for the grace period.  A job will continue to run (and maybe finish) while the preemption/deletion is being processed.

Reply:
Ok. So it should only be in the event of delete_done and not called_delete or preempted?

Reply:
Also, see my comment above about the order of events not mattering with the way I set up the trigger.

Reply:
If we only care about delete_done events, then I can revert the EntryIterator back to how it was before where it looked at entries in descending order until it reached the mark.

Reply:
> A job will continue to run (and maybe finish) while the preemption/deletion is being processed.

See my comment about left normalizing the time points. I think we'll handle this case correctly the way the code is currently.

Reply:
I just saw your comment above about billing up until the preemption event. I'm going to keep my code the way it is.